### ISSUE TITLE
Review ThreadLocal usage in context of JDK 21's virtual threads

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#15701]
jOOQ contains a few internal `ThreadLocal` instances, which are mainly required due to shortcomings of JDBC (e.g. UDTs can't really serialise / deserialise themselves to ojdbc without some way of accessing a JDBC `Connection`, which the JDBC `SQLData` SPIs don't give access to`).

With JDK 21's virtual threads, some of these `ThreadLocal` usages may be outdated, so we could perhaps use scoped variables instead, or something else.

See also:
- https://twitter.com/mariofusco/status/1709914958916387224?s=20
- https://github.com/jOOQ/jOOQ/issues/15177

**Comment by victornoel** [comment]
Hey @lukaseder, great to see this is on your todo list :)

I wanted to mention that it could also be interesting to revisit the use of `synchronized` as it creates "thread pinning" for now with Java 21. The usual alternative is to use locks from the standard lib. Also, you are maybe aware of it, but in a future Java version, `synchronized` will work out of the box, so it could be argued that we should just wait :P

Cheers

**Comment by lukaseder** [comment]
> I wanted to mention that it could also be interesting to revisit the use of `synchronized` as it creates "thread pinning" for now with Java 21. The usual alternative is to use locks from the standard lib. Also, you are maybe aware of it, but in a future Java version, `synchronized` will work out of the box, so it could be argued that we should just wait :P

I'll definitely wait. `synchronized` is much simpler than any lock API usage.

**Comment by victornoel** [comment]
Yep, makes sense, and I believe that in most case it is harmless to pin the thread if there is no blocking operations happening inside the synchronized so most certainly this is not going to prevent using jOOQ with Virtual Threads :)

**Comment by SightStudio** [comment]
**'JEP 491: Synchronize Virtual Threads without Pinning'** will be shipped in jdk 24


https://openjdk.org/jeps/491
